🐄 🛸 Update implementation of HPO suggestion

cthoyt opened the pull request:
This PR does the following:

1. Replaces the custom implementation of a default discrete integer distribution with the one added recently in Optuna (I think based on my suggestion!). This gets us free handling of questionable bounds/step size combinations that's built in to Optuna.
2. Generalizes the power_two scale to allow arbitrary powers to be set.
3. Adds the `log` option to linear integer search
4. Updates the default negative sampler range to use the `log` option

cthoyt committed:
Update implementation of HPO suggestion

Trigger CI

cthoyt committed:
Update hpo.py

Trigger CI

mberr commented on src/pykeen/sampling/negative_sampler.py:
```diff
@@ -21,7 +21,9 @@ class NegativeSampler(ABC):
     #: The default strategy for optimizing the negative sampler's hyper-parameters
-    hpo_default: ClassVar[Mapping[str, Mapping[str, Any]]]
+    hpo_default: ClassVar[Mapping[str, Mapping[str, Any]]] = dict(
+        num_negs_per_pos=dict(type=int, low=0, high=2, scale='power', base=10),
```
This means we only have three different `num_neg_per_pos` values by default, right?

mberr replied:
https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial.suggest_int

with `log=True` seems to be more appropriate to me. It still explores different scales uniformly, but does not "lose" possible `num_neg_per_pos` values

cthoyt replied:
True, but are there real meaningful differences in exploring the difference between 2 and 3, or 50 and 60? I like the current solution as a way of significantly reducing the possible space to explore.

We could go either way, though. I guess we should bake in the ability to use log anyway

cthoyt replied:
Updated in 2440c5b2

cthoyt committed:
Enable specifying log scale

cthoyt committed:
Remove negative sampler update

Now that there are some tests for this, I think we can merge this PR without passing judgement on this one yet

Trigger CI

cthoyt committed:
Update negative_sampler.py

Trigger CI
